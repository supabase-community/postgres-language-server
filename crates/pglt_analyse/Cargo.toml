
[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pglt_analyse"
repository.workspace = true
version              = "0.0.0"


[dependencies]
pglt_console.workspace      = true
pglt_diagnostics.workspace  = true
pglt_query_ext.workspace    = true
pglt_schema_cache.workspace = true
rustc-hash                  = { workspace = true }

biome_deserialize        = { workspace = true, optional = true }
biome_deserialize_macros = { workspace = true, optional = true }
enumflags2.workspace     = true
schemars                 = { workspace = true, optional = true }
serde                    = { workspace = true, features = ["derive"], optional = true }
text-size.workspace      = true

[features]
serde = ["dep:serde", "dep:schemars", "dep:biome_deserialize", "dep:biome_deserialize_macros"]
