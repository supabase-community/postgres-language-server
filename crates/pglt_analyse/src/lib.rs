mod categories;
pub mod context;
mod filter;
pub mod macros;
pub mod options;
mod registry;
mod rule;

// Re-exported for use in the `declare_group` macro
pub use pglt_diagnostics::category_concat;

pub use crate::categories::{
    ActionCategory, RefactorKind, RuleCategories, RuleCategoriesBuilder, RuleCategory,
    SUPPRESSION_ACTION_CATEGORY, SourceActionKind,
};
pub use crate::filter::{AnalysisFilter, GroupKey, RuleFilter, RuleKey};
pub use crate::options::{AnalyserOptions, AnalyserRules};
pub use crate::registry::{
    MetadataRegistry, RegistryRuleParams, RegistryVisitor, RuleRegistry, RuleRegistryBuilder,
};
pub use crate::rule::{
    GroupCategory, Rule, RuleDiagnostic, RuleGroup, RuleMeta, RuleMetadata, RuleSource,
};
