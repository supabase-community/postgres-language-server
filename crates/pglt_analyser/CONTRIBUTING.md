# Analyser

## Creating a rule

When creating or updating a lint rule, you need to be aware that there's a lot of generated code inside our toolchain.
Our CI ensures that this code is not out of sync and fails otherwise.
See the [code generation section](#code-generation) for more details.

To create a new rule, you have to create and update several files.
Because it is a bit tedious, we provide an easy way to create and test your rule using [Just](https://just.systems/man/en/).
_Just_ is not part of the rust toolchain, you have to install it with [a package manager](https://just.systems/man/en/chapter_4.html).

### Choose a name

We follow a naming convention according to what the rule does:

1. Forbid a concept

   ```block
   ban<Concept>
   ```

   When a rule's sole intention is to **forbid a single concept** the rule should be named using the `ban` prefix.

   Example: "banDropColumn"

2. Mandate a concept

   ```block
   use<Concept>
   ```

   When a rule's sole intention is to **mandate a single concept** the rule should be named using the `use` prefix.

### Explain a rule to the user

A rule should be informative to the user, and give as much explanation as possible.

When writing a rule, you must adhere to the following **pillars**:

1. Explain to the user the error. Generally, this is the message of the diagnostic.
2. Explain to the user **why** the error is triggered. Generally, this is implemented with an additional node.
3. Tell the user what they should do. Generally, this is implemented using a code action. If a code action is not applicable a note should tell the user what they should do to fix the error.

### Create and implement the rule

> [!TIP]
> As a developer, you aren't forced to make a rule perfect in one PR. Instead, you are encouraged to lay out a plan and to split the work into multiple PRs.
>
> If you aren't familiar with the APIs, this is an option that you have. If you decide to use this option, you should make sure to describe your plan in an issue.

Let's say we want to create a new **lint** rule called `useMyRuleName`, follow these steps:

1. Run the command

   ```shell
   just new-lintrule safety useMyRuleName
   ```

   The script will generate a bunch of files inside the `pglt_analyser` crate.
   Among the other files, you'll find a file called `use_my_new_rule_name.rs` inside the `pglt_analyser/lib/src/lint/safety` folder. You'll implement your rule in this file.

1. The `Options` type doesn't have to be used, so it can be considered optional. However, it has to be defined as `type Options = ()`.
1. Implement the `run` function: The function is called for every statement, and should return zero or more diagnostics. Follow the [pillars](#explain-a-rule-to-the-user) when writing the message of a diagnostic

Don't forget to format your code with `just f` and lint with `just l`.

That's it! Now, let's test the rule.

### Rule configuration

Some rules may allow customization using options.
We try to keep rule options to a minimum and only when needed.
Before adding an option, it's worth a discussion.

Let's assume that the rule we implement support the following options:

- `behavior`: a string among `"A"`, `"B"`, and `"C"`;
- `threshold`: an integer between 0 and 255;
- `behaviorExceptions`: an array of strings.

We would like to set the options in the `pglsp.toml` configuration file:

```toml
[linter.rules.safety.myRule]
level = "warn"
options = {
    behavior = "A"
    threshold = 20
    behaviorExceptions = ["one", "two"]
}
```

The first step is to create the Rust data representation of the rule's options.

```rust
#[derive(Clone, Debug, Default)]
pub struct MyRuleOptions {
    behavior: Behavior,
    threshold: u8,
    behavior_exceptions: Box<[Box<str>]>
}

#[derive(Clone, Debug, Defaul)]
pub enum Behavior {
    #[default]
    A,
    B,
    C,
}
```

Note that we use a boxed slice `Box<[Box<str>]>` instead of `Vec<String>`.
This allows saving memory: [boxed slices and boxed str use two instead of three words](https://nnethercote.github.io/perf-book/type-sizes.html#boxed-slices).

With these types in place, you can set the associated type `Options` of the rule:

```rust
impl Rule for MyRule {
    type Options = MyRuleOptions;
}
```

A rule can retrieve its options with:

```rust
let options = ctx.options();
```

The compiler should warn you that `MyRuleOptions` does not implement some required types.
We currently require implementing _serde_'s traits `Deserialize`/`Serialize`.

Also, we use other `serde` macros to adjust the JSON configuration:

- `rename_all = "snake_case"`: it renames all fields in camel-case, so they are in line with the naming style of the `pglsp.toml`.
- `deny_unknown_fields`: it raises an error if the configuration contains extraneous fields.
- `default`: it uses the `Default` value when the field is missing from `pglsp.toml`. This macro makes the field optional.

You can simply use a derive macros:

```rust
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(rename_all = "snake_case", deny_unknown_fields, default)]
pub struct MyRuleOptions {
    #[serde(default, skip_serializing_if = "is_default")]
    main_behavior: Behavior,

    #[serde(default, skip_serializing_if = "is_default")]
    extra_behaviors: Vec<Behavior>,
}

#[derive(Debug, Default, Clone)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub enum Behavior {
    #[default]
    A,
    B,
    C,
}
```

### Coding the rule

Below, there are many tips and guidelines on how to create a lint rule using our infrastructure.

#### `declare_lint_rule`

This macro is used to declare an analyzer rule type, and implement the [RuleMeta] trait for it.

The macro itself expects the following syntax:

```rust
use pglt_analyse::declare_lint_rule;

declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        recommended: false,
    }
}
```

##### Lint rules inspired by other lint rules

If a **lint** rule is inspired by an existing rule from other ecosystems (Squawk etc.), you can add a new metadata to the macro called `source`. Its value is `&'static [RuleSource]`, which is a reference to a slice of `RuleSource` elements, each representing a different source.

If you're implementing a lint rule that matches the behaviour of the Squawk rule `ban-drop-column`, you'll use the variant `::Squawk` and pass the name of the rule:

```rust
use pglt_analyse::{declare_lint_rule, RuleSource};

declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        recommended: false,
        sources: &[RuleSource::Squawk("ban-drop-column")],
    }
}
```

#### Category Macro

Declaring a rule using `declare_lint_rule!` will cause a new `rule_category!`
macro to be declared in the surrounding module. This macro can be used to
refer to the corresponding diagnostic category for this lint rule, if it
has one. Using this macro instead of getting the category for a diagnostic
by dynamically parsing its string name has the advantage of statically
injecting the category at compile time and checking that it is correctly
registered to the `pglt_diagnostics` library.

```rust
declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        recommended: false,
    }
}

impl Rule for BanDropColumn {
    type Options = Options;

    fn run(ctx: &RuleContext<Self>) -> Vec<RuleDiagnostic> {
        vec![RuleDiagnostic::new(
            rule_category!(),
            None,
            "message",
        )]
    }
}
```

### Document the rule

The documentation needs to adhere to the following rules:

- The **first** paragraph of the documentation is used as brief description of the rule, and it **must** be written in one single line. Breaking the paragraph in multiple lines will break the table content of the rules page.
- The next paragraphs can be used to further document the rule with as many details as you see fit.
- The documentation must have a `## Examples` header, followed by two headers: `### Invalid` and `### Valid`. `### Invalid` must go first because we need to show when the rule is triggered.
- Rule options if any, must be documented in the `## Options` section.
- Each code block must have `sql` set as language defined.
- When adding _invalid_ snippets in the `### Invalid` section, you must use the `expect_diagnostic` code block property. We use this property to generate a diagnostic and attach it to the snippet. A snippet **must emit only ONE diagnostic**.
- When adding _valid_ snippets in the `### Valid` section, you can use one single snippet.
- You can use the code block property `ignore` to tell the code generation script to **not generate a diagnostic for an invalid snippet**.

Here's an example of how the documentation could look like:

````rust
declare_lint_rule! {
    /// Dropping a column may break existing clients.
    ///
    /// Update your application code to no longer read or write the column.
    ///
    /// You can leave the column as nullable or delete the column once queries no longer select or modify the column.
    ///
    /// ## Examples
    ///
    /// ### Invalid
    ///
    /// ```sql,expect_diagnostic
    /// alter table test drop column id;
    /// ```
    ///
    pub BanDropColumn {
        version: "next",
        name: "banDropColumn",
        recommended: true,
        sources: &[RuleSource::Squawk("ban-drop-column")],
    }
}
````

This will cause the documentation generator to ensure the rule does emit
exactly one diagnostic for this code, and to include a snapshot for the
diagnostic in the resulting documentation page.

### Testing the Rule

#### Quick Test

To quickly test your rule, head to the `pglt_analyser/src/lib.rs` file and modify the `debug_test` function.

You should:

- remove the `#[ignore]` macro if present
- change the content of the `SQL` static `&str` to whatever you need
- pass your group and rule to the `RuleFilter::Rule(..)`

If you run the test, you'll see any diagnostics your rule created in your console.

### Code generation

For simplicity, use `just` to run all the commands with:

```shell
just gen-lint
```

### Commit your work

Once the rule implemented, tested, and documented, you are ready to open a pull request!

Stage and commit your changes:

```shell
> git add -A
> git commit -m 'feat(pglt_analyser): myRuleName'
```

### Deprecate a rule

There are occasions when a rule must be deprecated, to avoid breaking changes. The reason
of deprecation can be multiple.

In order to do, the macro allows adding additional field to add the reason for deprecation

````rust
use pglt_analyse::declare_lint_rule;

declare_lint_rule! {
    /// Dropping a column may break existing clients.
    ///
    /// Update your application code to no longer read or write the column.
    ///
    /// You can leave the column as nullable or delete the column once queries no longer select or modify the column.
    ///
    /// ## Examples
    ///
    /// ### Invalid
    ///
    /// ```sql,expect_diagnostic
    /// alter table test drop column id;
    /// ```
    ///
    pub BanDropColumn {
        version: "next",
        name: "banDropColumn",
        recommended: true,
        deprecated: true,
        sources: &[RuleSource::Squawk("ban-drop-column")],
    }
}
````
