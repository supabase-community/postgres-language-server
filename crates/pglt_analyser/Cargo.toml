
[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pglt_analyser"
repository.workspace = true
version              = "0.0.0"

[dependencies]
pglt_analyse   = { workspace = true }
pglt_console   = { workspace = true }
pglt_query_ext = { workspace = true }
serde          = { workspace = true }
