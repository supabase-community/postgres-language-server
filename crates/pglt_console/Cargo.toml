[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pglt_console"
repository.workspace = true
version              = "0.0.0"


[dependencies]
pglt_markup = { workspace = true }
text-size   = { workspace = true }

schemars             = { workspace = true, optional = true }
serde                = { workspace = true, optional = true, features = ["derive"] }
termcolor            = { workspace = true }
unicode-segmentation = "1.12.0"
unicode-width        = { workspace = true }

[dev-dependencies]
trybuild = "1.0.99"

[features]
serde_markup = ["serde", "schemars"]

[lib]
doctest = false
