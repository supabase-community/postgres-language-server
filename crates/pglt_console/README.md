# `pglt_console`

The crate contains a general abstraction over printing messages (formatted with markup) and diagnostics to a console.

## Acknowledgement

This crate was initially forked from [biome](https://github.com/biomejs/biome).

