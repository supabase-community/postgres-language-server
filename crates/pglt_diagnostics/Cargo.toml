[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pglt_diagnostics"
repository.workspace = true
version              = "0.0.0"


[dependencies]
backtrace                   = "0.3.74"
bpaf                        = { workspace = true }
enumflags2                  = { workspace = true }
pglt_console                = { workspace = true, features = ["serde_markup"] }
pglt_diagnostics_categories = { workspace = true, features = ["serde"] }
pglt_diagnostics_macros     = { workspace = true }
pglt_text_edit              = { workspace = true }
schemars                    = { workspace = true, optional = true }
serde                       = { workspace = true, features = ["derive"] }
serde_json                  = { workspace = true }
termcolor                   = { workspace = true }
text-size.workspace         = true
unicode-width               = { workspace = true }

[features]
schema = ["schemars", "pglt_text_edit/schemars", "pglt_diagnostics_categories/schemars"]

[dev-dependencies]

[lib]
doctest = false
