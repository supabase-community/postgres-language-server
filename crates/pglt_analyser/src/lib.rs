use std::{ops::Deref, sync::LazyLock};

use pglt_analyse::{
    AnalyserOptions, AnalysisFilter, MetadataRegistry, RegistryRuleParams, RuleDiagnostic,
    RuleRegistry,
};
pub use registry::visit_registry;

mod lint;
pub mod options;
mod registry;

pub static METADATA: LazyLock<MetadataRegistry> = LazyLock::new(|| {
    let mut metadata = MetadataRegistry::default();
    visit_registry(&mut metadata);
    metadata
});

/// Main entry point to the analyser.
pub struct Analyser<'a> {
    /// Holds the metadata for all the rules statically known to the analyser
    /// we need this later when we add suppression support
    #[allow(dead_code)]
    metadata: &'a MetadataRegistry,

    /// Holds all rule options
    options: &'a AnalyserOptions,

    /// Holds all rules
    registry: RuleRegistry,
}

pub struct AnalyserContext<'a> {
    pub root: &'a pglt_query_ext::NodeEnum,
}

pub struct AnalyserConfig<'a> {
    pub options: &'a AnalyserOptions,
    pub filter: AnalysisFilter<'a>,
}

impl<'a> Analyser<'a> {
    pub fn new(conf: AnalyserConfig<'a>) -> Self {
        let mut builder = RuleRegistry::builder(&conf.filter);
        visit_registry(&mut builder);
        let registry = builder.build();

        Self {
            metadata: METADATA.deref(),
            registry,
            options: conf.options,
        }
    }

    pub fn run(&self, ctx: AnalyserContext) -> Vec<RuleDiagnostic> {
        let params = RegistryRuleParams {
            root: ctx.root,
            options: self.options,
        };

        self.registry
            .rules
            .iter()
            .flat_map(|rule| (rule.run)(&params))
            .collect::<Vec<_>>()
    }
}

#[cfg(test)]
mod tests {
    use core::slice;

    use pglt_analyse::{AnalyserOptions, AnalysisFilter, RuleFilter};
    use pglt_console::{
        fmt::{Formatter, Termcolor},
        markup, Markup,
    };
    use pglt_diagnostics::PrintDiagnostic;
    use termcolor::NoColor;

    use crate::Analyser;

    #[ignore]
    #[test]
    fn debug_test() {
        fn markup_to_string(markup: Markup) -> String {
            let mut buffer = Vec::new();
            let mut write = Termcolor(NoColor::new(&mut buffer));
            let mut fmt = Formatter::new(&mut write);
            fmt.write_markup(markup).unwrap();

            String::from_utf8(buffer).unwrap()
        }

        const SQL: &str = r#"alter table test drop column id;"#;
        let rule_filter = RuleFilter::Rule("safety", "banDropColumn");

        let filter = AnalysisFilter {
            enabled_rules: Some(slice::from_ref(&rule_filter)),
            ..Default::default()
        };

        let ast = pglt_query_ext::parse(SQL).expect("failed to parse SQL");

        let options = AnalyserOptions::default();

        let analyser = Analyser::new(crate::AnalyserConfig {
            options: &options,
            filter,
        });

        let results = analyser.run(crate::AnalyserContext { root: &ast });

        println!("*******************");
        for result in &results {
            let text = markup_to_string(markup! {
                {PrintDiagnostic::simple(result)}
            });
            eprintln!("{}", text);
        }
        println!("*******************");

        // assert_eq!(results, vec![]);
    }
}
