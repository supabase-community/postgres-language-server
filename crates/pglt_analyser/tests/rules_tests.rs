use core::slice;
use std::{fmt::Write, fs::read_to_string, path::Path};

use pglt_analyse::{AnalyserOptions, AnalysisFilter, RuleDiagnostic, RuleFilter};
use pglt_analyser::{Analyser, AnalyserConfig, AnalyserContext};
use pglt_console::StdDisplay;
use pglt_diagnostics::PrintDiagnostic;

pglt_test_macros::gen_tests! {
  "tests/specs/**/*.sql",
  crate::rule_test
}

fn rule_test(full_path: &'static str, _: &str, _: &str) {
    let input_file = Path::new(full_path);

    let (group, rule, fname) = parse_test_path(input_file);

    let rule_filter = RuleFilter::Rule(group.as_str(), rule.as_str());
    let filter = AnalysisFilter {
        enabled_rules: Some(slice::from_ref(&rule_filter)),
        ..Default::default()
    };

    let query =
        read_to_string(full_path).expect(format!("Failed to read file: {} ", full_path).as_str());

    let ast = pglt_query_ext::parse(&query).expect("failed to parse SQL");
    let options = AnalyserOptions::default();
    let analyser = Analyser::new(AnalyserConfig {
        options: &options,
        filter,
    });

    let results = analyser.run(AnalyserContext { root: &ast });

    let mut snapshot = String::new();
    write_snapshot(&mut snapshot, query.as_str(), results.as_slice());

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(fname, snapshot);
    });

    let expectation = Expectation::from_file(&query);
    expectation.assert(results.as_slice());
}

fn parse_test_path(path: &Path) -> (String, String, String) {
    let mut comps: Vec<&str> = path
        .components()
        .into_iter()
        .map(|c| c.as_os_str().to_str().unwrap())
        .collect();

    let fname = comps.pop().unwrap();
    let rule = comps.pop().unwrap();
    let group = comps.pop().unwrap();

    (group.into(), rule.into(), fname.into())
}

fn write_snapshot(snapshot: &mut String, query: &str, diagnostics: &[RuleDiagnostic]) {
    writeln!(snapshot, "# Input").unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot, "{query}").unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot).unwrap();

    if !diagnostics.is_empty() {
        writeln!(snapshot, "# Diagnostics").unwrap();
        for diagnostic in diagnostics {
            let printer = PrintDiagnostic::simple(diagnostic);

            writeln!(snapshot, "{}", StdDisplay(printer)).unwrap();
            writeln!(snapshot).unwrap();
        }
    }
}

enum Expectation {
    NoDiagnostics,
    AnyDiagnostics,
    OnlyOne(String),
}

impl Expectation {
    fn from_file(content: &str) -> Self {
        for line in content.lines() {
            if line.contains("expect_no_diagnostics") {
                return Self::NoDiagnostics;
            }

            if line.contains("expect_only_") {
                let kind = line
                    .splitn(3, "_")
                    .last()
                    .expect("Use pattern: `-- expect_only_<category>`")
                    .trim();

                return Self::OnlyOne(kind.into());
            }
        }

        Self::AnyDiagnostics
    }

    fn assert(&self, diagnostics: &[RuleDiagnostic]) {
        match self {
            Self::NoDiagnostics => {
                if !diagnostics.is_empty() {
                    panic!("This test should not have any diagnostics.");
                }
            }
            Self::OnlyOne(category) => {
                let found_kinds = diagnostics
                    .iter()
                    .map(|d| d.get_category_name())
                    .collect::<Vec<&str>>()
                    .join(", ");

                if diagnostics.len() != 1 || diagnostics[0].get_category_name() != category {
                    panic!(
                        "This test should only have one diagnostic of kind: {category}\nReceived: {found_kinds}"
                    );
                }
            }
            Self::AnyDiagnostics => {}
        }
    }
}
