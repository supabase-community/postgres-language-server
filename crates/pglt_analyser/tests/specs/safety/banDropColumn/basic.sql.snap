---
source: crates/pglt_analyser/tests/rules_tests.rs
expression: snapshot
---
# Input
```
-- expect_only_lint/safety/banDropColumn
alter table test
drop column id;
```

# Diagnostics
lint/safety/banDropColumn ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Dropping a column may break existing clients.
  
  i You can leave the column as nullable or delete the column once queries no longer select or modify the column.
